### ISSUE TITLE
Question about usage without polluting the build files

### ISSUE THREAD

**Issue by hedefalk** [sbt/sbt-sriracha#6]
Hi Eugene,

Just found out about this thing which might be a solution for the problem I've felt I've had for ages when trying to work with many small modules which is my preferred way of keeping stuff tidy.

Background:

Back in the day, I was using Maven and independent non-monorepo micromodules. Typically, for a monolith webapp I could have more than 20 different pom based modules and working with them simultanously in an Eclipse workspace.

m2eclipse, the eclipse plugin for Maven had this really nice thing that it could handle the classpath resolution so that if you had A depending on B-SNAPSHOT and you actually had B-SNAPSHOT in your workspace, it would automatically be a source dependency instead of a managed maven dependency. 

This probably 15 years ago and it was really powerful. There was nice ui for it too, since you'd have the vcs url in your pom, you could just start working on A and if you noticed you needed to fix something in B, you would just right-click the dependency and something like "materialize as source project" and it would `git checkout` and fix it all up within seconds.

I've been missing this thing ever since starting with sbt and I never really felt anyone knew what I was talking about.

Lately I've been doing a lot of work in node and there you have `npm link` which basically solves the same problem in a very smooth way.

But now I'm back in Scala land and I'm working with this module A depending on B. I need to iterate over changes in B that I use in A and I need to first change the version of B in A/build.sbt to -SNAPSHOT and then I need to iterate the whole cycle of

* B> publishLocal
* A> ;reload ;clean ;~test:compile

over and over again.

That's the backstory. And this project really looks like it could solve some of my issues. I however feel my hands are a bit tied in that I wouldn't want to mess up this clients build definitions too much with my workflow specific things.

Therefore I'm wondering if there's a smooth way to use a separate .sbt file not under vcs that could inject these settings non-intrusively…?

So if I could leave the A/build.sbt like:

     "org.myorg" %% "b" % "1.0",

but then have a viktor.sbt file gitignore that switched this out to:

     ProjectRef(IO.toURI("~" / "projects" / "b"), "org.myorg" %% "b" % "1.0")

My sbt foo is still not good enough to tell if this is possible?

Anyway, thanks for this project! (and a bunch of other too :) )

**Comment by eed3si9n** [comment]
This is an interesting constraint. Thanks for taking your time to post this! Here's the user story I am hearing:

- I want to git clone some random project A from GitHub, and start hacking on its transitive dependency B without messing up the original `build.sbt` of A.

Does the capture the idea? I think we can do this if we mimicked the way subproject dependencies work today, instead of just using it like sriracha is doing right now. To satisfy the user story I am picturing a new `srirachaResolver` that gets added to all subprojects if the plugin is enabled. When the dependency resolution asks "hey do you have `"org.myorg" %% "b" % "1.0"`" it would ignore the version part and say yes.

**Comment by hedefalk** [comment]
>I want to git clone some random project A from GitHub, and start hacking on its transitive dependency B without messing up the original build.sbt of A.

Yes, that's definitely the gist! Sorry about the essay version of it :)

You think it would be hard to implement this?

**Comment by eed3si9n** [comment]
The resolver itself might not be that hard, but it also needs everything else subproject dependency takes for granted, like making sure it becomes part of the build, and that `compile` would trigger transitive compilation. All that would be some trial-and-error work.

**Comment by hedefalk** [comment]
Just found your blogpost from two years ago and also the comment from OlegYch:

```
object BuildUtil {
private val localLibs = Try(IO.read(file("local-libs.txt"))).map(_.split('\n').map(_.trim).toSet).getOrElse(Set())
implicit class ProjectExt(p: Project) {
/**
* if artifactId is in local-libs.txt then add project at a sibling directory with that name as a dependency, otherwise add that artifact as a regular dependency
*/
def addAcmeLib(artifactId: String, version: String) = {
val org = "com.acme"
if (localLibs(artifactId)) {
val ref = ProjectRef(file("..") / artifactId, artifactId)
val exclude = allDependencies ~= (_.map(_.exclude(org, artifactId)))
p.dependsOn(ref).aggregate(ref).settings(exclude)
} else {
p.settings(libraryDependencies += org %% artifactId % version)
}}}}
```

If `allDependencies ~= (_.map(_.exclude(org, artifactId)))` thing works fine, maybe this is all I need to switch it out without needing to change the build.sbt?

**Comment by hedefalk** [comment]
Hm, don't know what I meant with that last comment… I guess the idea looks like its just the same as sriracha.

I'm revisiting this now since we're working with modularizing very fine grain libs and would like to temporarily work with them like a monorepo when figuring out the architecture while still handling them in separate git repos and isolated versioning+publishing.

If I understand the problem, it's that a plugin can only inject settings like `allDependencies` but cannot stuff into `dependsOn` because this is not a setting. So therefore we're using enriched methods like `def sourceDependency(ref: ClasspathDep[ProjectReference], bin: ModuleID)` on the project definition instead. But the problem is then that I cannot externalize this from the vcs'd build definition…

So I would like by some kind of external definitions on .gitignore to:

- [x] Remove the binary dependency - seems easy enough with allDependencies ~= (_.map(_.exclude………
- [ ] Add the source dependency 

I'm looking into the API - is there no way that could be done with `extraProjects`, maybe by an extra indirection? Or will I hit something like https://github.com/sbt/sbt/issues/4976 then?

Is a resolver the only reasonable option?

**Comment by hedefalk** [comment]
I just want to clarify my "feature request" because I realize I've been rambling.

I think sriracha does everything I want, except I cannot use it without polluting the original build files. There seems to be no way I can use sriracha or the approach it uses by adding .gitignore'd extra files which I would be fine with. 

I've been trying to experiment myself, but it seems the culprit is just that `.dependsOn` isn't callable from a plugin the way I need - I don't have a reference to the Project.

>The resolver itself might not be that hard, but it also needs everything else subproject dependency takes for granted, like making sure it becomes part of the build, and that compile would trigger transitive compilation. 

when doing `.dependsOn` like sriracha does, then everything is taken care of right? So just some way to do that from "outside" is really all I'm looking for…
